File: frontend/src/domain/host/model/HostRequest.ts
export enum HostBindingType {
    HTTP = "HTTP",
    HTTPS = "HTTPS",
}

export enum HostRouteType {
    PROXY = "PROXY",
    REDIRECT = "REDIRECT",
    STATIC_RESPONSE = "STATIC_RESPONSE",
    INTEGRATION = "INTEGRATION",
    EXECUTE_CODE = "EXECUTE_CODE",
    STATIC_FILES = "STATIC_FILES",
}

export enum HostRouteSourceCodeLanguage {
    JAVASCRIPT = "JAVASCRIPT",
    LUA = "LUA",
}

export interface HostFeatureSet {
    websocketsSupport: boolean
    http2Support: boolean
    redirectHttpToHttps: boolean
}

export interface HostBinding {
    type: HostBindingType
    ip: string
    port: number
    certificateId?: string
}

export interface HostRouteStaticResponse {
    statusCode: number
    payload?: string
    headers?: Record<string, string>
}

export interface HostRouteSettings {
    includeForwardHeaders: boolean
    proxySslServerName: boolean
    keepOriginalDomainName: boolean
    directoryListingEnabled: boolean
    custom?: string
}

export interface HostRouteSourceCode {
    language: HostRouteSourceCodeLanguage
    code: string
    mainFunction?: string
}

export interface HostRoute {
    priority: number
    enabled: boolean
    type: HostRouteType
    sourcePath: string
    settings: HostRouteSettings
    targetUri?: string
    response?: HostRouteStaticResponse
    integration?: HostRouteIntegration
    accessListId?: string
    redirectCode?: number
    sourceCode?: HostRouteSourceCode
}

export interface HostRouteIntegration {
    integrationId: string
    optionId: string
}

export default interface HostRequest {
    enabled: boolean
    defaultServer: boolean
    useGlobalBindings: boolean
    domainNames?: string[]
    routes: HostRoute[]
    bindings: HostBinding[]
    featureSet: HostFeatureSet
    accessListId?: string
}

File: api/src/main/kotlin/br/com/dillmann/nginxignition/api/host/model/HostRequest.kt
package br.com.dillmann.nginxignition.api.host.model

import br.com.dillmann.nginxignition.api.common.serialization.UuidString
import br.com.dillmann.nginxignition.core.host.Host
import kotlinx.serialization.Serializable

@Serializable
internal data class HostRequest(
    val enabled: Boolean,
    val defaultServer: Boolean,
    val useGlobalBindings: Boolean,
    val domainNames: List<String> = emptyList(),
    val routes: List<Route> = emptyList(),
    val bindings: List<Binding> = emptyList(),
    val featureSet: FeatureSet,
    val accessListId: UuidString? = null,
) {
    @Serializable
    data class Route (
        val priority: Int,
        val type: Host.RouteType,
        val sourcePath: String,
        val settings: RouteSettings,
        val targetUri: String? = null,
        val redirectCode: Int? = null,
        val response: StaticResponse? = null,
        val integration: IntegrationConfig? = null,
        val accessListId: UuidString? = null,
        val sourceCode: RouteSourceCode? = null,
    )

    @Serializable
    data class RouteSourceCode(
        val language: Host.SourceCodeLanguage,
        val code: String,
        val mainFunction: String? = null,
    )

    @Serializable
    data class RouteSettings (
        val includeForwardHeaders: Boolean,
        val proxySslServerName: Boolean,
        val keepOriginalDomainName: Boolean,
        val custom: String? = null,
    )

    @Serializable
    data class IntegrationConfig(
        val integrationId: String,
        val optionId: String,
    )

    @Serializable
    data class StaticResponse(
        val statusCode: Int,
        val payload: String? = null,
        val headers: Map<String, String> = emptyMap(),
    )

    @Serializable
    data class FeatureSet(
        val websocketsSupport: Boolean,
        val http2Support: Boolean,
        val redirectHttpToHttps: Boolean,
    )

    @Serializable
    data class Binding(
        val type: Host.BindingType,
        val ip: String,
        val port: Int,
        val certificateId: UuidString? = null,
    )
}
